File: sims/dashboard/lib/telemetry.ts
import type { TelemetryEvent, TelemetrySummary, TelemetryToolStat } from "./types";

export function parseTelemetry(jsonl: string): TelemetryEvent[] {
  const out: TelemetryEvent[] = [];
  for (const line of jsonl.split("\n")) {
    const s = line.trim();
    if (!s) continue;
    let m: any;
    try {
      m = JSON.parse(s);
    } catch {
      continue;
    }
    out.push({
      ts: str(m.ts),
      sessionId: str(m.session_id),
      tool: str(m.tool),
      args: m.args && typeof m.args === "object" ? m.args : {},
      resultCount: typeof m.result_count === "number" ? m.result_count : 0,
      isError: m.is_error === true,
      error: str(m.error),
      latencyMs: typeof m.latency_ms === "number" ? m.latency_ms : 0,
    });
  }
  return out;
}

export function summarizeTelemetry(events: TelemetryEvent[]): TelemetrySummary {
  const byTool = new Map<string, TelemetryToolStat>();
  const latencies: number[] = [];
  const queries = new Set<string>();
  const docs = new Set<string>();
  let errors = 0;

  for (const e of events) {
    const st = byTool.get(e.tool) ?? { tool: e.tool, calls: 0, errors: 0 };
    st.calls++;
    if (e.isError) {
      st.errors++;
      errors++;
    }
    byTool.set(e.tool, st);
    latencies.push(e.latencyMs);
    if (e.tool === "search_fiskaly_docs" && typeof e.args.query === "string") queries.add(e.args.query);
    if (e.tool === "fetch_fiskaly_doc" && typeof e.args.id === "string") docs.add(e.args.id);
  }

  latencies.sort((a, b) => a - b);
  return {
    total: events.length,
    errors,
    byTool: [...byTool.values()].sort((a, b) => b.calls - a.calls),
    p50LatencyMs: percentile(latencies, 50),
    p95LatencyMs: percentile(latencies, 95),
    queries: [...queries],
    docsFetched: [...docs],
  };
}

function percentile(sorted: number[], p: number): number {
  if (sorted.length === 0) return 0;
  const idx = Math.min(sorted.length - 1, Math.floor((p / 100) * sorted.length));
  return sorted[idx];
}

function str(v: any): string {
  return typeof v === "string" ? v : "";
}

File: eval-harness/backend/internal/artifacts/telemetry.go
package artifacts

import (
	"cmp"
	"encoding/json"
	"slices"
)

// ParseTelemetry parses a JSONL telemetry file into typed events.
func ParseTelemetry(jsonl string) []TelemetryEvent {
	out := []TelemetryEvent{}
	scanJSONL(jsonl, func(r map[string]json.RawMessage) {
		e := TelemetryEvent{Args: map[string]any{}}
		if v, ok := r["ts"]; ok {
			_ = json.Unmarshal(v, &e.Ts)
		}
		if v, ok := r["tool"]; ok {
			_ = json.Unmarshal(v, &e.Tool)
		}
		if v, ok := r["args"]; ok {
			var args map[string]any
			if err := json.Unmarshal(v, &args); err == nil {
				e.Args = args
			}
		}
		if v, ok := r["result_count"]; ok {
			_ = json.Unmarshal(v, &e.ResultCount)
		}
		if v, ok := r["is_error"]; ok {
			_ = json.Unmarshal(v, &e.IsError)
		}
		if v, ok := r["error"]; ok {
			_ = json.Unmarshal(v, &e.Error)
		}
		if v, ok := r["latency_ms"]; ok {
			_ = json.Unmarshal(v, &e.LatencyMs)
		}
		out = append(out, e)
	})
	return out
}

// SummarizeTelemetry aggregates telemetry events into a summary.
func SummarizeTelemetry(events []TelemetryEvent) TelemetrySummary {
	byTool := map[string]*TelemetryToolStat{}
	toolOrder := []string{}
	latencies := make([]int, 0, len(events))
	queriesSeen := map[string]bool{}
	var queries []string
	docsSeen := map[string]bool{}
	var docs []string
	errors := 0

	for _, e := range events {
		st, exists := byTool[e.Tool]
		if !exists {
			st = &TelemetryToolStat{Tool: e.Tool}
			byTool[e.Tool] = st
			toolOrder = append(toolOrder, e.Tool)
		}
		st.Calls++
		if e.IsError {
			st.Errors++
			errors++
		}
		latencies = append(latencies, e.LatencyMs)
		if e.Tool == "search_fiskaly_docs" {
			if q, ok := e.Args["query"].(string); ok && !queriesSeen[q] {
				queriesSeen[q] = true
				queries = append(queries, q)
			}
		}
		if e.Tool == "fetch_fiskaly_doc" {
			if id, ok := e.Args["id"].(string); ok && !docsSeen[id] {
				docsSeen[id] = true
				docs = append(docs, id)
			}
		}
	}

	sorted := make([]TelemetryToolStat, 0, len(toolOrder))
	for _, t := range toolOrder {
		sorted = append(sorted, *byTool[t])
	}
	slices.SortStableFunc(sorted, func(a, b TelemetryToolStat) int {
		return cmp.Compare(b.Calls, a.Calls)
	})

	slices.Sort(latencies)

	if queries == nil {
		queries = []string{}
	}
	if docs == nil {
		docs = []string{}
	}

	return TelemetrySummary{
		Total:        len(events),
		Errors:       errors,
		ByTool:       sorted,
		P50LatencyMs: percentile(latencies, 50),
		P95LatencyMs: percentile(latencies, 95),
		Queries:      queries,
		DocsFetched:  docs,
	}
}

func percentile(sorted []int, p int) int {
	if len(sorted) == 0 {
		return 0
	}
	return sorted[min(len(sorted)-1, p*len(sorted)/100)]
}
